therapeutic/models/nh_clinical_therapeutic_level.py

```diff
@@ -0,0 +1,182 @@
+
+        Unfortunately Odoo does not send the value of read-only fields to the
+        server so records are created with `False` for their frequency. This
+        override is necessary to ensure the correct frequency is in the
+        database and can therefore be used by other features.
+
+        :param values:
+        :return:
+        """
+        if 'frequency' not in values or not values['frequency']:
+            values['frequency'] = 60
+        return super(NhClinicalPatientObservationTherapeuticLevel, self)\
+            .create(values)
+
+    @api.onchange('level')
+    def _set_fields_based_on_level(self):
+        """
+        Currently frequency should always be 60 for all levels except level 2.
+        This is enforced in the UI by making the frequency field read-only.
+
+        This onchange method ensures that the read-only frequency field shows
+        'Every Hour' when levels 1, 3, or 4 are selected.
+        """
+        if self.is_level(1):
+            self.frequency = 60
+            self.staff_to_patient_ratio = False
+        elif self.is_level(2):
+            self.frequency = False
+            self.staff_to_patient_ratio = False
+        elif self.is_level(3) or self.is_level(4):
+            self.frequency = 60
+
+    def default_get(self, cr, uid, fields, context=None):
+        """
+        Ensure that the patient field is pre-populated on the form view when
+        creating a new level record from the wardboard view.
+
+        :param cr:
+        :param uid:
+        :param fields:
+        :param context:
+        :return:
+        """
+        field_defaults_dict = \
+            super(NhClinicalPatientObservationTherapeuticLevel, self)\
+            .default_get(cr, uid, fields, context=None)
+
+        if 'active_id' in context \
+                and 'active_model' in context \
+                and context['active_model'] == 'nh.clinical.wardboard':
+            wardboard_model = self.pool['nh.clinical.wardboard']
+            wardboard = wardboard_model.browse(cr, uid, context['active_id'])
+            patient_id = wardboard.patient_id.id
+
+            field_defaults_dict['patient'] = patient_id
+        return field_defaults_dict
+
+    def save(self, cr, uid, ids, context=None):
+        """
+        There is a 'Save' button in the 'Set Therapeutic Obs' dialog accessible
+        from the patient form. This method exists purely to stop that button
+        from blowing up.
+
+        Pressing the button creates the therapeutic level
+        record even without calling anything (I think the `oe_form_button_save`
+        class on the button triggers an action in Odoo's JavaScript to do this)
+        but if there is nothing set on the button to call then it blows up
+        during module load so this method exists as a hack purely to stop that
+        blow up.
+
+        :param cr:
+        :param uid:
+        :param ids:
+        :param context:
+        :return:
+        """
+        pass
```

Comment:
I don't understand this. Why is this blowing up?

Reply:
When you create a button you've got to give it a method to call otherwise it blows up. Thing is I don't need the button to call any method because it is handling the saving automatically for me somehow, I suspect the JS is doing it because it when it sees the `oe_form_button_save` class on the button. I create this method so I could give the button something to call that just does nothing. Means I still get the automatic save and it doesn't blow up.

Reply:
I don't know why Odoo blows up when you don't give the button a method to call though.

Reply:
So if your suspicions are correct, then the JS is looking at the class of the button being there, that means the changes could be getting saved all the time? Including when the popup is closed? Or when the levels are changed? Or are you changing the class of the button when the button is pressed, so then the save occurs? 
The save should only happen when the button is pressed, so that's the method the button should be calling.

Reply:
I should have been more specific, I suspect that the JS event handler for the button press checks the class of the button and if it sees the class then saves it for you without you having to call a method yourself.

I've checked locally and the record is only saved when the button is actually pressed.

Reply:
I can add an E2E test for this real quick, I should have added it already tbh.

Reply:
Oh I see, ok :) yes please.

Reply:
I've added [the test](https://github.com/bjss/BJSS_liveobs_automation/pull/22/commits/5f9d1fd3136e24d6de30b1d25b25bf91f13d56f8) and it is passing locally.

Reply:
Oops missed a step, [added now](https://github.com/bjss/BJSS_liveobs_automation/pull/22).